### PR TITLE
Refactor ABNF rules for DIDs and DID References.

### PR DESCRIPTION
**Pull request by dmitrizagidulin**
This PR integrates and reflects the current consensus on the DID ABNF rules, as well as DID Reference and Service Endpoint ABNF rules.

Specifically:

* Separates the DID Reference ABNF from the DID ABNF
* Differentiates a DID fragment from a Service fragment
* Specifies ABNF rules for did query, did path
* Clarifies use cases for DID References (service endpoints resolution, key references, etc). 
* Updates service endpoint `id:` properties in examples.

Replaces PR #106, addresses #85.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c-ccg/did-spec/pull/168.html" title="Last updated on Feb 16, 2019, 6:54 PM UTC (89a234c)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c-ccg/did-spec/168/aeb7bc6...89a234c.html" title="Last updated on Feb 16, 2019, 6:54 PM UTC (89a234c)">Diff</a>

### CODE REVIEW

**Comment by dmitrizagidulin** [comment]
ABNF rules refactored. What remains (this should probably go in a separate PR? I'm not sure):

* Add an explanation for what DID Reference Service Paths are _for_
* Link to the corresponding section in the DID Resolution spec.

File: index.html [hunk 1]
```diff
@@ -715,67 +716,85 @@ <h2>
...
-<a href="#the-generic-did-scheme"></a>) is identical to a URI
-fragment and MUST conform to the ABNF of the fragment ABNF rule in
-[[RFC3986]]. A DID fragment MUST be used only as a method-independent
-pointer into the DID Document to identify a unique key description or
-other DID Document component. To resolve this pointer, the complete
-DID reference including the DID fragment MUST be used as the value of
-the id key for the target JSON object.
-      </p>
+        <p>
+  A generic <a>DID fragment</a> (the <pre>did-fragment</pre> rule in Section
```
Comment (davidlehn) [inline index.html]: I think all the uses of `<pre>` in sentences in this patch should be using `<code>`?  The `<pre>` rendering is wrong.

File: index.html [hunk 2]
```diff
@@ -715,67 +716,85 @@ <h2>
       </p>
```
Comment (davidlehn) [inline index.html]: As long as ABNF is being worked on, it seems weird to have the above loosely defining `ALPHA`.  If we're going to define `ALPHA` why not just put it below?  `DIGIT` should get the same treatment either way.  I'm not sure what the best practice is for this. Could maybe say:
```
The following is the ABNF definition using the syntax in [[RFC5234]]
which defines <code>ALPHA</code> and <code>DIGIT</code>.
```

File: index.html (hunk 1 above)
Comment (dmitrizagidulin) [reply index.html]: Whoops.. Fixing..

File: index.html [hunk 3]
```diff
@@ -715,67 +716,85 @@ <h2>
...
-Paths
+DID Reference
       </h2>
 
-      <p>
-A generic <a>DID path</a> (the did-path rule in Section <a href="#the-generic-did-scheme"></a>) is identical to a URI path and MUST
-conform to the ABNF of the path-rootless ABNF rule in [[RFC3986]]. A
-DID path SHOULD be used to address resources available via a DID
-service endpoint. See Section <a href="#service-endpoints"></a>.
+      <p>DID References are used for two main purposes: to refer to sections
```
Comment (davidlehn) [inline index.html]: style: start text on next line flush left.

Comment (dmitrizagidulin) [reply index.html]: Thx, fixed.

File: index.html [hunk 4]
```diff
@@ -800,7 +820,7 @@ <h2>
 
         <li>
 Case sensitivity and normalization of the value of the
-specific-idstring rule in Section <a href="#the-generic-did-scheme">
+<code>method-specific-idstring</code> rule in Section <a href="#the-generic-did-scheme">
```
Comment (davidlehn) [inline index.html]: Remove the newline inside `<a ...></a>`.

Comment (dmitrizagidulin) [reply index.html]: Thx, fixed.

File: index.html (hunk 1 above)
Comment (dmitrizagidulin) [reply index.html]: Fixed, thanks.

File: index.html (hunk 2 above)
Comment (dmitrizagidulin) [reply index.html]: Thanks! Changed to your suggestion.

File: index.html [hunk 5]
```diff
@@ -534,6 +534,7 @@ <h2>
     "publicKeyPem": "-----BEGIN PUBLIC KEY...END PUBLIC KEY-----\r\n"
   }],
   "service": [{
+    "id": "#service123",
```
Comment (davidlehn) [inline index.html]: use full URI

File: index.html [hunk 6]
```diff
@@ -1402,31 +1423,31 @@ <h2>
       <pre class="example nohighlight" title="Various service endpoints">
 {
   "service": [{
-    "id": "did:example:123456789abcdefghi;openid",
+    "id": "#openid",
```
Comment (davidlehn) [inline index.html]: use full URI (here and below)

File: index.html (hunk 5 above)
Comment (dmitrizagidulin) [reply index.html]: Fixed, thanks!

File: index.html (hunk 6 above)
Comment (dmitrizagidulin) [reply index.html]: Fixed, thanks!

**Comment by mwherman2000** [comment]
How soon can this PR be accepted?

File: index.html [hunk 7]
```diff
@@ -702,80 +703,100 @@ <h2>
...
-      <p>
-A specific DID scheme MAY specify ABNF rules for DID paths that are
-more restrictive than the generic rules in this section.
-      </p>
-    </section>
+      <pre class="nohighlight">
+did-reference      = did (did-fragment-reference / did-service-reference)
+did-fragment-reference  = "#" did-fragment
+did-service-reference   = ";" service-id [ "/" service-path ] [ "?" service-query ] [ "#" service-fragment ]
+service-id         = ALPHA / DIGIT / "." / "-" / "_"
```
Comment (davidlehn) [inline index.html]: The following are not defined: `did-fragment`, `service-path`, `service-query`, `service-fragment`.

File: index.html [hunk 8]
```diff
@@ -702,80 +703,100 @@ <h2>
...
+
+Additionally, a DID can be used as a component in a DID Reference, described in
+the <a href="#did-reference">DID Reference section</a> below.
       </p>
 
       <p>
-Following is the ABNF definition using the syntax in [[RFC5234]]
-        (which defines ALPHA as upper or lowercase A-Z).
+The following is the ABNF definition using the syntax in [[RFC5234]]
+which defines <code>ALPHA</code> and <code>DIGIT</code>.
```
Comment (davidlehn) [inline index.html]: Should this be moved up a section and phrased so that it covers The Generic DID Scheme section and DID Reference section?

**Comment by davidlehn** [comment]
@dmitrizagidulin If the preview link is correct, there's a `<section>` tag issue somewhere.  Everything is under section 4, vs the current version where that is not the case.

**Comment by davidlehn** [comment]
@dmitrizagidulin Does `terms.html` needs some updating too to match these changes?  For instance, the metions and links about "DID Path" .  (And while fixing that, maybe reword so the hard coded easy breakable "section x.y" text in that file isn't used)

File: index.html [hunk 9]
```diff
@@ -702,80 +703,100 @@ <h2>
...
       </h2>
 
       <p>
-A generic <a>DID path</a> (the did-path rule in Section <a href="#the-generic-did-scheme"></a>) is identical to a URI path and MUST
-conform to the ABNF of the path-rootless ABNF rule in [[RFC3986]]. A
-DID path SHOULD be used to address resources available via a DID
-service endpoint. See Section <a href="#service-endpoints"></a>.
+DID References are used for two main purposes: to refer to sections
+inside a DID Document (via <a href="#did-fragments">DID Fragments</a>), and as a
+portable way to refer to Service Endpoint resources, queries and fragments.
```
Comment (davidlehn) [inline index.html]: oxford comma and link:
```suggestion
portable way to refer to <a>Service Endpoint</a> resources, queries, and fragments.
```

File: index.html [hunk 10]
```diff
@@ -702,80 +703,100 @@ <h2>
...
+  given by the target URL without needing to rely on graph-based processing.
+        </p>
+        <p>
+  Implementations SHOULD NOT prevent the use of JSON pointers ([[RFC6901]]).
+        </p>
+      </section>
+
+      <section>
+        <h3>
+      DID Reference Service Paths, Queries and Fragments
```
Comment (davidlehn) [inline index.html]:
```suggestion
      DID Reference Service Paths, Queries, and Fragments
```

**Comment by davidlehn** [comment]
There are more places where the def `<a>term</a>` style could be used. Unsure how far that sort of linking needs to go.

File: index.html (hunk 7 above)
Comment (dmitrizagidulin) [reply index.html]: So, that's intentional. `did-fragment`, `-path` and `-query` are mentioned in the next paragraph (basically, the ABNF for them is the same as in the URI RFC).

Comment (davidlehn) [reply index.html]: Ah, I see.  Unsure if that needs to be noted out or not.  Leave it until someone else gets confused too.

Comment (dmitrizagidulin) [reply index.html]: Heh, yeah, I was wondering that same thing, but wasn't sure how to make it clearer. Does ABNF syntax have comments? :)

File: index.html [hunk 11]
```diff
@@ -702,80 +703,100 @@ <h2>
...
+  fragment and MUST conform to the ABNF of the fragment ABNF rule in
+  [[RFC3986]]. A DID fragment MUST be used only as a method-independent
+  pointer into the DID Document to identify a unique key description or
+  other DID Document component. To resolve this pointer, the complete
+  DID reference including the DID fragment MUST be used as the value of
+  the id key for the target JSON object.
+        </p>
+
+        <p>
+  A specific DID scheme MAY specify ABNF rules for DID fragments that
```
Comment (davidlehn) [inline index.html]: Should this be `A specific <a>DID method</a> MAY...`?

File: index.html [hunk 12]
```diff
@@ -702,80 +703,100 @@ <h2>
...
+rule in [[RFC3986]]. The path is used to address resources available via a DID
+service endpoint. See Section <a href="#service-endpoints"></a>.
       </p>
 
       <p>
-It is desirable that we enable tree-based processing of DIDs that include
-DID fragments (which resolve directly within the DID document) to locate
-metadata contained directly in the DID document or the service resource
-given by the target URL without needing to rely on graph-based processing.
+A specific DID scheme MAY specify ABNF rules for DID Reference service paths
```
Comment (davidlehn) [inline index.html]: Should this be `A specific <a>DID method</a> MAY`?

File: index.html [hunk 13]
```diff
@@ -702,80 +703,100 @@ <h2>
...
+
+      <section>
+        <h3>
+      DID Reference Service Paths, Queries, and Fragments
+        </h3>
 
       <p>
-A specific DID scheme MAY specify ABNF rules for DID fragments that
-are more restrictive than the generic rules in this section.
+A DID Reference Service path (the <code>service-path</code> rule above) is
```
Comment (davidlehn) [inline index.html]: To be consistent:
```suggestion
A DID Reference service path (the <code>service-path</code> rule in Section
<a href="#did-reference"></a>) is
```

File: index.html [hunk 14]
```diff
@@ -702,80 +703,100 @@ <h2>
...
-DID fragments (which resolve directly within the DID document) to locate
-metadata contained directly in the DID document or the service resource
-given by the target URL without needing to rely on graph-based processing.
+A specific DID scheme MAY specify ABNF rules for DID Reference service paths
+that are more restrictive than the generic rules in this section.
       </p>
+
       <p>
-Implementations SHOULD NOT prevent the use of JSON pointers ([[RFC6901]]).
+A DID Reference Service <code>query</code> and <code>fragment</code> parts are
```
Comment (davidlehn) [inline index.html]: To be consistent:
```suggestion
The DID Reference service query and service fragment
(the <code>service-query</code> and <code>service-fragment</code> rules
in Section <a href="#did-reference"></a>) parts are
```

File: index.html (hunk 12 above)
Comment (dmitrizagidulin) [reply index.html]: Probably, yeah

File: index.html (hunk 7 above)
Comment (davidlehn) [reply index.html]: The ABNF RFC is linked a few lines up!  Comments: https://tools.ietf.org/html/rfc5234#section-3.9

**Comment by dhh1128** [comment]
I do not understand why we feel the need to use a different delimiter to distinguish service fragments from other types of fragments. Why can't we just use / for all fragments of any type, and require that their ID properties be unique within the DID Document?

**Comment by mwherman2000** [comment]
@dhh1128 I agree.  At the `did-url` grammar level, [we should focus on allowable syntax constructs - in support for the desired  capabilities we want `did-url`s to have](https://hyperonomy.com/2019/03/11/giving-grammars-written-with-abnf-notation-the-respect-they-deserve/) ...not try to encode all sorts of specific capabilities and functions into the syntax.

Examples:
1. IMO a bad use of syntax: https://github.com/WebOfTrustInfo/rwot8-barcelona/pull/164#issuecomment-470869225
2. IMO a better use of syntax: https://github.com/w3c-ccg/did-resolution/issues/32#issuecomment-471179669

The latter, at a syntax level, supports many levels of functional/capability innovation without having to change the syntax.

**Comment by peacekeeper** [comment]
@dmitrizagidulin do you think everything in this PR has been addressed by https://github.com/w3c-ccg/did-spec/pull/189 (in fact that one even cherry-picked a few commits from here)? Or would you say this PR here still has content that needs to be merged?

**Comment by kdenhartog** [comment]
I wanted to highlight the example of what I brought up on the DID call today about key-type matrix param being used as a method to specify a specific key-type within a grouping of key-types. For example, let's say in my DID-Doc I specify a grouping of `RSAVerificationKey2016` `RSAVerificationKey2017` and  `RSAVerificationKey2018` and call this grouping `RSAVerificationSuite` which is what the key-type would be listed in my DID Doc.

E.g. my DID Doc would be:

```
{
  "@context": ["https://w3id.org/did/v1", "https://w3id.org/security/v1"],
  "id": "did:example:123456789abcdefghi",
  ...
  "publicKey": [{
    "id": "did:example:123456789abcdefghi#keys-1",
    "type": "RsaVerificationSuite",
    "controller": "did:example:123456789abcdefghi",
    "publicKeyPem": "-----BEGIN PUBLIC KEY...END PUBLIC KEY-----\r\n"
  }],
  ...
}
```

Then when referencing to the public key for verification, I could use `did:example:123456789abcdefghi;key-type=RSAVerification2018#keys-1`

To suggest that the same key can be used with multiple key types.

If you were to use this for service-type it would be something like a grouping of services where a common set of services sit behind a single endpoint. So for example, I may run an Agent which supports DIDComm v1 and DIDCommv2 and I would specify the serialization of the object by using:

`did:example:123456789abcdefghi;key-type=DIDComm-JSON#agent`

or 

`did:example:123456789abcdefghi;key-type=DIDComm-CBOR#agent`

This is just an early example of where I think it may be used. I'm not sure if this indirection is useful, necessary, or neither. With that said, I want to take the time to agree with @dmitrizagidulin point raised that matrix params are not a method of optimizing resolver implementations. They should be used to identify a specific resource when the DID Doc contains metadata that describes a set.

**Comment by dmitrizagidulin** [comment]
This PR is superceded by PR #189, so, it's ok to close this one.
